Update simpleion.load function to support iterating infinite streams

alimg opened the pull request:
*Issue #, if available:* N/A

*Description of changes:* Converted the ionc_read implementation to return and iterator instance to allow processing large ION streams without having to load entire input buffers into memory. 

Sharing this pull request to see if there's interest to see these changes merged. If so, I can work on few improvements to this changeset to add more integration tests and verify there are no memory leak with valgrind etc..

*Testing:* Tested with unit test and made sure they all pass. And used this version briefly in a separate workload to process some big files.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

tgregg commented:
@alimg This looks like a great idea to me; yes, we have interest in seeing these changes merged. We will work on scheduling time to review this PR in detail.

cheqianh reviewed (commented):
This is a really good change, Thanks for the contribution!

&nbsp;
For your comment:
> Sharing this pull request to see if there's interest to see these changes merged. If so, I can work on few improvements to this changeset to add more integration tests and verify there are no memory leak with valgrind etc..

Yes we are interested to see this PR got merged. Could you add some integration tests and essential unit tests to make sure the iterator functions correctly? 

In addition, I did some pre-profiling 
1. I did some performance measurement and it's almost as fast as the previous C extension, which is good.
2.  I did some memory leak profiling (I know you might do profiling too, I just write it down here to keep this on track). It seems that the `ionc_read` method has some memory leak issue when `parse_eagerly` equals to both true and false. So I think the leak might happen somewhere in the iterator's setup/next methods other than ion_c parse level. Could you please take a look at this?

alimg commented:
Updated the PR to fix a leak in dealloc method and some some other leaks that were revealed mostly around exceptional cases. I used `PYTHONMALLOC=malloc valgrind --show-leak-kinds=definite --leak-check=full  ` --addopts "-s tests/test_simpleion.py "` command-line to check for memory leaks. Full test suite looks mostly clean now but there are still some illegal memory accesses and small sources of leaks. I'll try to take a stab at them another time.

alimg commented on amazon/ion/ioncmodule.c:
```diff
@@ -1047,7 +1103,7 @@ static iERR ionc_read_timestamp(hREADER hreader, PyObject** timestamp_out) {
                 }
 
-                char* dec_num = malloc(fractional_precision + 2);
+                char* dec_num = malloc(fractional_precision + 6); // TODO: decQuadToString call below seems to overflow the allocated memory
```
Valgrind revealed some illegal memory access made by the 3rd party code where it tried to write beyond the number of bytes we allocated here. I've bumped up the padding a bit, but still not sure if this is the right way

alimg replied:
Dumped some info from my environment [here](https://gist.github.com/alimg/632440698bcf8949f8b6fc8552de9670) but I must admit I don't know what could be platform dependent here.

Is there a reason not to use [DECQUAD_String](https://github.com/amzn/ion-c/blob/a541a810b8e0471262f7bfda04283859a07b7d10/decNumber/example7.c#L18) here? I've updated this to match the other examples used in the ion-c code base, but I hope I didn't remove any specific memory saving trick put in place before.

cheqianh replied:
Yeah I agree, using `DECQUAD_String` is a good idea. Change `+2` to `+6` is not a big deal too. I'm just curious dump what data causes the illegal memory access.

cheqianh reviewed (commented):
Thanks for the contribution! I only had some minor questions and left comments below.

I left some comments about documentation and tests, please feel free to change them if you would like to. Otherwise I would also be happy to create a pull request later to add little more tests and description.

For two [memory leaks](https://github.com/alimg/ion-c/commit/f8add5ed1bfee0deec63926af94bad65fd6f6d29) found in ion-c, I saw a similar issue opened earlier and it's fixed in [this PR](https://github.com/amzn/ion-c/pull/257), I'll take a look at the other one.

I added a [commit](https://github.com/amzn/ion-python/pull/172/commits/1c3e0002f710b4b0a7cfcd14e5b6cf97f0e95cc7) to fix the conflict due to the recent change in c-extension-beta branch.

alimg commented on amazon/ion/ioncmodule.c:
```diff
@@ -1474,46 +1532,158 @@ iERR ionc_read_all(hREADER hreader, PyObject* container, BOOL in_struct, BOOL em
+}
+
+PyObject* ionc_read_iter(PyObject *self) {
+    ionc_read_Iterator *iterator = (ionc_read_Iterator*) self;
```
no reason at all, I meant to delete this line

alimg commented on tests/test_simpleion.py:
```diff
@@ -612,6 +612,24 @@ def test_roundtrip(p):
+    out = BytesIO()
+    dump(obj, out, binary=is_binary, indent=indent, tuple_as_sexp=tuple_as_sexp)
+    out.seek(0)
+    res = load(out, single_value=False, parse_eagerly=True)
```
Yes, I added this to make sure the new code paths on the Python side are tested. 

When `parse_eagerly is True`, lazy mode always returns an iterator, even if there was only one ion value in the input stream. I wanted to capture this in the tests.

alimg commented:
>     3. I added a [commit](https://github.com/amzn/ion-python/pull/172/commits/1c3e0002f710b4b0a7cfcd14e5b6cf97f0e95cc7) to fix the conflict due to recent change in the c-extension-beta branch.
Rebased from the upstream to address this.

cheqianh commented:
Thanks for the rebase. You might not have to rebase it everytime because we can use `Squash and merge` button below to squash all commits into one later.
